### ISSUE TITLE
Spurious slowdown in `ImGui_ImplOpenGL3_RenderDrawData` function when creating and joining threads

### ISSUE THREAD

**Issue by ArktisDev** [ocornut/imgui#4148]
**Version/Branch of Dear ImGui:**

Version: 1.82
Branch: tagged release

**Back-end/Renderer/Compiler/OS**

Back-ends: imgui_impl_glfw.cpp + imgui_impl_opengl3.cpp
Compiler: MinGW g++ 10.2.0
Operating System: Windows

**My Issue/Question:**

When creating threads inside of render loop, `ImGui_ImplOpenGL3_RenderDrawData` function experiences spurious slowdowns for seemingly no reason. Initially when I encountered this behavior, I expected this to be a problem with my application, but I've reproduced the behavior in the GLFW + OpenGL3 example file.

**Standalone, minimal, complete and verifiable example:**

Starting with the GLFW + OpenGL3 example file, add an include `#include <thread>`, and then add this code inside of the main render loop to create and immediately join a new thread every 60 frames.
```c++
static int counter = 0;
if (counter == 60) {
    counter = 0;
    std::thread th([] () {});
    th.join();
} else {
    counter++;
}
```

Full code with timing:

<details><summary>CLICK ME</summary>
<p>

```c++
// Dear ImGui: standalone example application for GLFW + OpenGL 3, using programmable pipeline
// (GLFW is a cross-platform general purpose library for handling windows, inputs, OpenGL/Vulkan/Metal graphics context creation, etc.)
// If you are new to Dear ImGui, read documentation from the docs/ folder + read the top of imgui.cpp.
// Read online: https://github.com/ocornut/imgui/tree/master/docs

#include "imgui/imgui.h"
#include "imgui/imgui_impl_glfw.h"
#include "imgui/imgui_impl_opengl3.h"
#include <stdio.h>
#include <thread>
#include <chrono>
#include <iostream>

#if defined(IMGUI_IMPL_OPENGL_ES2)
#include <GLES2/gl2.h>
// About Desktop OpenGL function loaders:
//  Modern desktop OpenGL doesn't have a standard portable header file to load OpenGL function pointers.
//  Helper libraries are often used for this purpose! Here we are supporting a few common ones (gl3w, glew, glad).
//  You may use another loader/header of your choice (glext, glLoadGen, etc.), or chose to manually implement your own.
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GL3W)
#include <GL/gl3w.h>            // Initialize with gl3wInit()
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLEW)
#include <GL/glew.h>            // Initialize with glewInit()
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLAD)
#include <glad/glad.h>          // Initialize with gladLoadGL()
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLAD2)
#include <glad/gl.h>            // Initialize with gladLoadGL(...) or gladLoaderLoadGL()
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLBINDING2)
#define GLFW_INCLUDE_NONE       // GLFW including OpenGL headers causes ambiguity or multiple definition errors.
#include <glbinding/Binding.h>  // Initialize with glbinding::Binding::initialize()
#include <glbinding/gl/gl.h>
using namespace gl;
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLBINDING3)
#define GLFW_INCLUDE_NONE       // GLFW including OpenGL headers causes ambiguity or multiple definition errors.
#include <glbinding/glbinding.h>// Initialize with glbinding::initialize()
#include <glbinding/gl/gl.h>
using namespace gl;
#else
#include IMGUI_IMPL_OPENGL_LOADER_CUSTOM
#endif

// Include glfw3.h after our OpenGL definitions
#include <GLFW/glfw3.h>

// [Win32] Our example includes a copy of glfw3.lib pre-compiled with VS2010 to maximize ease of testing and compatibility with old VS compilers.
// To link with VS2010-era libraries, VS2015+ requires linking with legacy_stdio_definitions.lib, which we do using this pragma.
// Your own project should not be affected, as you are likely to link with a newer binary of GLFW that is adequate for your version of Visual Studio.
#if defined(_MSC_VER) && (_MSC_VER >= 1900) && !defined(IMGUI_DISABLE_WIN32_FUNCTIONS)
#pragma comment(lib, "legacy_stdio_definitions")
#endif

static void glfw_error_callback(int error, const char* description)
{
    fprintf(stderr, "Glfw Error %d: %s\n", error, description);
}

int main(int, char**)
{
    // Setup window
    glfwSetErrorCallback(glfw_error_callback);
    if (!glfwInit())
        return 1;

    // Decide GL+GLSL versions
#if defined(IMGUI_IMPL_OPENGL_ES2)
    // GL ES 2.0 + GLSL 100
    const char* glsl_version = "#version 100";
    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 2);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0);
    glfwWindowHint(GLFW_CLIENT_API, GLFW_OPENGL_ES_API);
#elif defined(__APPLE__)
    // GL 3.2 + GLSL 150
    const char* glsl_version = "#version 150";
    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 2);
    glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);  // 3.2+ only
    glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);            // Required on Mac
#else
    // GL 3.0 + GLSL 130
    const char* glsl_version = "#version 130";
    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0);
    //glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);  // 3.2+ only
    //glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);            // 3.0+ only
#endif

    // Create window with graphics context
    GLFWwindow* window = glfwCreateWindow(1280, 720, "Dear ImGui GLFW+OpenGL3 example", NULL, NULL);
    if (window == NULL)
        return 1;
    glfwMakeContextCurrent(window);
    glfwSwapInterval(1); // Enable vsync

    // Initialize OpenGL loader
#if defined(IMGUI_IMPL_OPENGL_LOADER_GL3W)
    bool err = gl3wInit() != 0;
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLEW)
    bool err = glewInit() != GLEW_OK;
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLAD)
    bool err = gladLoadGL() == 0;
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLAD2)
    bool err = gladLoadGL(glfwGetProcAddress) == 0; // glad2 recommend using the windowing library loader instead of the (optionally) bundled one.
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLBINDING2)
    bool err = false;
    glbinding::Binding::initialize();
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLBINDING3)
    bool err = false;
    glbinding::initialize([](const char* name) { return (glbinding::ProcAddress)glfwGetProcAddress(name); });
#else
    bool err = false; // If you use IMGUI_IMPL_OPENGL_LOADER_CUSTOM, your loader is likely to requires some form of initialization.
#endif
    if (err)
    {
        fprintf(stderr, "Failed to initialize OpenGL loader!\n");
        return 1;
    }

    // Setup Dear ImGui context
    IMGUI_CHECKVERSION();
    ImGui::CreateContext();
    ImGuiIO& io = ImGui::GetIO(); (void)io;
    //io.ConfigFlags |= ImGuiConfigFlags_NavEnableKeyboard;     // Enable Keyboard Controls
    //io.ConfigFlags |= ImGuiConfigFlags_NavEnableGamepad;      // Enable Gamepad Controls

    // Setup Dear ImGui style
    ImGui::StyleColorsDark();
    //ImGui::StyleColorsClassic();

    // Setup Platform/Renderer backends
    ImGui_ImplGlfw_InitForOpenGL(window, true);
    ImGui_ImplOpenGL3_Init(glsl_version);

    // Load Fonts
    // - If no fonts are loaded, dear imgui will use the default font. You can also load multiple fonts and use ImGui::PushFont()/PopFont() to select them.
    // - AddFontFromFileTTF() will return the ImFont* so you can store it if you need to select the font among multiple.
    // - If the file cannot be loaded, the function will return NULL. Please handle those errors in your application (e.g. use an assertion, or display an error and quit).
    // - The fonts will be rasterized at a given size (w/ oversampling) and stored into a texture when calling ImFontAtlas::Build()/GetTexDataAsXXXX(), which ImGui_ImplXXXX_NewFrame below will call.
    // - Read 'docs/FONTS.md' for more instructions and details.
    // - Remember that in C/C++ if you want to include a backslash \ in a string literal you need to write a double backslash \\ !
    //io.Fonts->AddFontDefault();
    //io.Fonts->AddFontFromFileTTF("../../misc/fonts/Roboto-Medium.ttf", 16.0f);
    //io.Fonts->AddFontFromFileTTF("../../misc/fonts/Cousine-Regular.ttf", 15.0f);
    //io.Fonts->AddFontFromFileTTF("../../misc/fonts/DroidSans.ttf", 16.0f);
    //io.Fonts->AddFontFromFileTTF("../../misc/fonts/ProggyTiny.ttf", 10.0f);
    //ImFont* font = io.Fonts->AddFontFromFileTTF("c:\\Windows\\Fonts\\ArialUni.ttf", 18.0f, NULL, io.Fonts->GetGlyphRangesJapanese());
    //IM_ASSERT(font != NULL);

    // Our state
    bool show_demo_window = true;
    bool show_another_window = false;
    ImVec4 clear_color = ImVec4(0.45f, 0.55f, 0.60f, 1.00f);
    
    std::thread th([] () {});
    // Main loop
    while (!glfwWindowShouldClose(window))
    {
        // Poll and handle events (inputs, window resize, etc.)
        // You can read the io.WantCaptureMouse, io.WantCaptureKeyboard flags to tell if dear imgui wants to use your inputs.
        // - When io.WantCaptureMouse is true, do not dispatch mouse input data to your main application.
        // - When io.WantCaptureKeyboard is true, do not dispatch keyboard input data to your main application.
        // Generally you may always pass all inputs to dear imgui, and hide them from your application based on those two flags.
        glfwPollEvents();
        
        static int counter = 0;
        if (counter == 60) {
            counter = 0;
            auto start = std::chrono::steady_clock::now();
            std::thread th([] () {});
            
            th.join();
            auto stop = std::chrono::steady_clock::now();
            std::cout << "threads: " << std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count() << std::endl;
        } else {
            counter++;
        }

        // Start the Dear ImGui frame
        ImGui_ImplOpenGL3_NewFrame();
        ImGui_ImplGlfw_NewFrame();
        ImGui::NewFrame();

        // 1. Show the big demo window (Most of the sample code is in ImGui::ShowDemoWindow()! You can browse its code to learn more about Dear ImGui!).
        if (show_demo_window)
            ImGui::ShowDemoWindow(&show_demo_window);

        // 2. Show a simple window that we create ourselves. We use a Begin/End pair to created a named window.
        {
            static float f = 0.0f;
            static int counter = 0;

            ImGui::Begin("Hello, world!");                          // Create a window called "Hello, world!" and append into it.

            ImGui::Text("This is some useful text.");               // Display some text (you can use a format strings too)
            ImGui::Checkbox("Demo Window", &show_demo_window);      // Edit bools storing our window open/close state
            ImGui::Checkbox("Another Window", &show_another_window);

            ImGui::SliderFloat("float", &f, 0.0f, 1.0f);            // Edit 1 float using a slider from 0.0f to 1.0f
            ImGui::ColorEdit3("clear color", (float*)&clear_color); // Edit 3 floats representing a color

            if (ImGui::Button("Button"))                            // Buttons return true when clicked (most widgets return true when edited/activated)
                counter++;
            ImGui::SameLine();
            ImGui::Text("counter = %d", counter);

            ImGui::Text("Application average %.3f ms/frame (%.1f FPS)", 1000.0f / ImGui::GetIO().Framerate, ImGui::GetIO().Framerate);
            ImGui::End();
        }

        // 3. Show another simple window.
        if (show_another_window)
        {
            ImGui::Begin("Another Window", &show_another_window);   // Pass a pointer to our bool variable (the window will have a closing button that will clear the bool when clicked)
            ImGui::Text("Hello from another window!");
            if (ImGui::Button("Close Me"))
                show_another_window = false;
            ImGui::End();
        }
        
        // Rendering
        ImGui::Render();
        int display_w, display_h;
        glfwGetFramebufferSize(window, &display_w, &display_h);
        glViewport(0, 0, display_w, display_h);
        glClearColor(clear_color.x * clear_color.w, clear_color.y * clear_color.w, clear_color.z * clear_color.w, clear_color.w);
        glClear(GL_COLOR_BUFFER_BIT);
        {
            auto start = std::chrono::steady_clock::now();
            
            ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
            
            auto stop = std::chrono::steady_clock::now();
            std::cout << "render: " << std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count() << std::endl;
        }
        

        glfwSwapBuffers(window);
    }
    
    th.join();

    // Cleanup
    ImGui_ImplOpenGL3_Shutdown();
    ImGui_ImplGlfw_Shutdown();
    ImGui::DestroyContext();

    glfwDestroyWindow(window);
    glfwTerminate();

    return 0;
}
```

</p>
</details>

**Observations**

To time the execution time of thread creation/joining and `ImGui_ImplOpenGL3_RenderDrawData` I used `std::chrono`.

Before adding thread creation and joining:
When I am creating and joining no threads in the main render loop, timing `ImGui_ImplOpenGL3_RenderDrawData` yields an average execution time of around 400 microseconds, well within an acceptable range.

After adding thread creation and joining:
Timing the execution time to create and join an empty thread like `std::thread th([] () {}); th.join()` in the render thread, I get an execution time of somewhere on the order of 200-300 microseconds. However, shortly after this thread joins, I see one spurious call to `ImGui_ImplOpenGL3_RenderDrawData` that doesn't take the standard 400 microseconds  to complete, instead it takes more than 40 milliseconds to complete, much slower than I need to be rendering frames.

If I increase the rate at which I generate and join threads, I see that every single time I am creating and joining a thread, I see this spurious slowdown either immediately or shortly after in the `ImGui_ImplOpenGL3_RenderDrawData` function.

**Comment by ocornut** [comment]
Thank you for posting such a detailed issue.

However, I'm not sure how you are expecting us to help you. That could be done to many factors and you'll need to investigate this further yourself in `ImGui_ImplOpenGL3_RenderDrawData()` to find answers but it's most likely to be due to interactions that are outside the scope of dear imgui and its backends. 

Feel free to investigate to post additional clues if you find them (will monitor subsequent posts), but until you can suggest/prove that there is something we can do on our end, I'm afraid we cannot be researching random issues like that.

**Comment by ArktisDev** [comment]
Thanks for the quick response!

After finding the issue, I changed my application to use a thread pool instead of generating new threads on the fly, since that doesn't cause this weird slowdown, so for now I've sort of avoided the issue. I believe thread pools are better coding practice for my specific use case anyways.

However, I am interested in figuring out what exactly is causing this issue, so when I get around to it sometime, I might investigate further what exactly is going wrong inside the `ImGui_ImplOpenGL3_RenderDrawData()` to see if it is any problem with ImGui.

**Comment by ocornut** [comment]
I would suggest using a profiler to try to find out if/why the main thread is stalling.

I guess OpenGL probably uses a TLS (thread-local-storage) variable to access its current context, meaning every OpenGL calls are going to access this TLS variable. However it is not expect that it would amount to a very meaningful performance cost.

**Comment by ArktisDev** [comment]
After some testing, I have found out some more about this issue, but not really a way to solve it, or any information about the exact behavior of OpenGL in this scenario.

**My Observations**
Timing the individual function calls in the `ImGui_ImplOpenGL3_RenderDrawData()` function, I've figured out that when the function is experiencing the slowdown it is occurring in the first call to `glGet...`. In the actual code, this function call:
```
GLenum last_active_texture; glGetIntegerv(GL_ACTIVE_TEXTURE, (GLint*)&last_active_texture);
```
is taking >95% of the execution time of `ImGui_ImplOpenGL3_RenderDrawData()` when it experiences a slowdown, while it takes only ~25% of the execution time in a normal function call. During normal calls to `ImGui_ImplOpenGL3_RenderDrawData()`, this call to `glGet...` is taking between 200 and 300 us. However on the calls which experience a slowdown, it is taking >95% of the time, so almost 40ms. It is [known ](https://www.khronos.org/opengl/wiki/Common_Mistakes#glGetFloatv_glGetBooleanv_glGetDoublev_glGetIntegerv) that this function is slow, so the ~25% of function execution time in normal cases is actually quite reasonable.

It seems that once a `glGet...` function is called once, subsequent calls do not experience this slowdown. Swapping this line in the `ImGui_ImplOpenGL3_RenderDrawData()` function:
```
GLenum last_active_texture; glGetIntegerv(GL_ACTIVE_TEXTURE, (GLint*)&last_active_texture);
```
with a line that is below it (and shouldn't impact anything about the function due to swapping the lines):
```
GLenum last_active_texture; glGetIntegerv(GL_ACTIVE_TEXTURE, (GLint*)&last_active_texture);
```
we see that the slowdown is now occuring in the new first call to `glGet...`

**Conclusions**
This is weird behavior. I would think that since I am not creating an OpenGL context in the new threads I am creating in the rendering loop, that OpenGL would not even react to them. In fact, I am not even doing anything in the new threads, I am just creating them with an empty lambda! Searching online, any sources I've found that discuss OpenGL and threads all discuss creating OpenGL contexts, or passing one around between threads, but that is not what I am doing here. Somehow it seems that OpenGL is recognizing that new threads are being created, and for some bizarre reason it is doing something about it, even though no OpenGL function calls, and in fact no function calls at all execute in these threads. This leads to the diagnosis:

**Diagnosis**
This is very likely not a problem that ImGui can solve. In order to do the rendering in the `ImGui_ImplOpenGL3_RenderDrawData()` function, you must store the OpenGL state before rendering, and restore it afterwards. Barring weird sparsely documented behavior of OpenGL when creating new threads unrelated to OpenGL (an unlikely scenario), this is most likely some sort of driver issue or something similar on my end. But thankfully it has a simple solution for me: just don't create new threads in the main render loop. It is usually better anyways to use a thread pool to manage worker threads like this. Since I've solved my issue for now, I'm not going to immediately try reupdating my drivers (everything else seems to work just fine), but if I need to revisit this behavior in the future, that should be one of the first tries at a solution.
